### ISSUE TITLE
NxN BlockMask / Cumulative Sequence Length

### ISSUE THREAD

**Issue by L-Reichardt** [meta-pytorch/attention-gym#187]
Hi,
I tried to implement FlexAttention for large batch training. Each attention layer computes attention within a window. My tensor is a batch-packed tensor to handle variable sequence lengths. The size of each batch sequence changes with the data (some batch samples are longer than other)

This means that I not only need to mask the window, but also the batches. In FlashAttention2 this is simple, i just pass a cumulative sequence length tensor, e.g.

bounds = torch.tensor([ 0, 1024, 2048, 3072, 4096, 5120, 6144, 7168, 8192, 9216, 10240, 11264, 12288, 13312, 14336, 15360, 16384, 17408, 18432, 19456, 20480, 21504, 22528, 23552, 24576, 25600, 26624, 27648, 28672, 29696, 30720, 31744, 32768, 33792, 34816, 35840, 36864, 37888, 38912, 39936, 40960, 41984, 43008, 44032, 45056, 46080, 47104, 48128, 49152, ... ]) # Here window size = 1024

The issue:
Im training with a very large batch size, so I have thousands of window/batch-sequence blocks.

I tried various ways to calculate the BlockMask, but it always materializes a NxN matrix. Using compile is extremely slow (~20x slower than FA2), since I have to recalculate the BlockMask at each sample. Padding is not an option, as this is too inefficient.

My question
Is there a way to use a similar cumulative sequence length tensor, as in FA2, without materializing memory-prohibitive NxN masks in any part of flex attention?v
